blog/2023-10-12-constructs-api-design-2.md

```diff
@@ -0,0 +1,316 @@
+  {
+    "id": "root/rose",
+    "kind": "rose",
+    "color": "red",
+    "nativeRegions": [
+      "Denmark"
+    ]
+  },
+  // ... rest of the garden data
+]
+```
+
+Now let's say we add the capability for users to _get_ the native regions of a flower.
+I'll also add a construct for representing a signpost in front of our garden.
+
+```js
+class Flower extends Construct {
+  get nativeRegions() {
+    return [...this._nativeRegions];
+  }
+
+  // ... rest of the class unchanged
+}
+
+class Signpost extends Construct {
+  constructor(scope, id, props) {
+    super(scope, id);
+    const allRegions = new Set(props.flowers.flatMap((f) => f.nativeRegions));
+
+    this._message = "Welcome to Tulip Trove, home to flowers from: ";
+    this._message += [...allRegions].join(", ");
+    this._message += ";";
+  }
+
+  toJson() {
+    return {
+      id: this.node.path,
+      message: this._message,
+    };
+  }
+}
+```
+
+Inside `Signpost`, I'm all of the native regions of the flowers passed to the signpost, de-duplicating them, and embedding them into a friendly message.
+
+Finally, I'll write some client code that tries using the signpost with some flowers:
+
+```js
+const garden = new Garden(undefined, "root");
+
+// add a flower
+const rose = new Flower(garden, "rose", { kind: "rose", color: "red" });
+rose.addNativeRegion("Denmark");
+
+// add a signpost
+new Signpost(garden, "signpost", { flowers: [rose] });
+
+// add more regions to our first flower
+rose.addNativeRegion("Turkey");
+rose.addNativeRegion("Greece");
+
+garden.synth();
+```
+
+When I synthesize my garden with `node garden.js`, I'm expecting the signpost to have a message like "Welcome to Tulip Trove, home to flowers from: Denmark, Turkey, Greece".
+But when I check `garden.json`, I find the **signpost message only mentions Denmark**:
+
+```json
+[
+  {
+    "id": "root/rose",
+    "kind": "rose",
+    "color": "red",
+    "nativeRegions": [
+      "Denmark",
+      "Turkey",
+      "Greece"
+    ]
+  },
+  {
+    "id": "root/signpost",
+    "message": "Welcome to Tuple Trove, home to flowers from: Denmark."
+  }
+]
+```
+
+Aw shucks.
+
+The problem, as you guessed, is that the state read by `Signpost` was stale.
+Since the signpost's message was calculated immediately, it wasn't changed when the rose's native regions was added to.
+
+But in some sense, it's not entirely `Signpost`'s fault - how was it supposed to know the field could change?
+It doesn't seem right that to have to look at the implementation of `Flower` in order to determine whether the data will be calculated later or not.
+We need a better way.
+
+## "Lazy" values to the rescue
+
+The approach we're going to take to solve this problem is to add support for a way of modeling values that aren't available yet, called `Lazy` values.
+
+Each constructs framework has a slightly different way of doing this, but the general idea is that instead of returning some state that could become stale, as we did here in `Flower`:
```

Comment:
```suggestion
Each construct's framework has a slightly different way of doing this, but the general idea is that instead of returning some state that could become stale, as we did here in `Flower`:
```

Reply:
```suggestion
Each construct framework has a slightly different way of doing this, but the general idea is that instead of returning some state that could become stale, as we did here in `Flower`:
```